### DISCUSSION TITLE
Hosting MCP Server?

### DISCUSSION THREAD

**Discussion by Jflick58** [modelcontextprotocol/modelcontextprotocol#1193] [Q&A - General]
### Pre-submission Checklist

- [X] I have checked that this question would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions and documentation for answers

### Question Category

- [ ] Protocol Specification
- [ ] SDK Usage
- [X] Server Implementation
- [X] General Implementation
- [ ] Documentation
- [ ] Other

### Your Question

I apologize if this is a dumb question. I really like the MCP architecture, but It appears that the core examples are all locally-ran servers, rather than networked. Specifically, I'm interested in adopting MCP as a mechanism to centrally host tools such as part of an enterprise agent stack. We have multiple potential clients that could consume it (e.g. Continue.dev, or a new internal-facing Typescript web client). Are there any examples yet of doing so? I see the transports documentation but I don't think I've seen any examples of actually integrating SSE transports with a client outside of the snippets.

**Comment by shenli** [comment]
Hi @Jflick58 I encoutered this issue. Currently there is no way to do this. See the discussion here: https://github.com/orgs/modelcontextprotocol/discussions/16 
So I use a proxy as a workaround: https://github.com/lightconetech/mcp-gateway  It is very easy to use.

**Reply by Logicoeur** [reply]
This is the way @Jflick58 . Until MCP support for remote servers is released you'll need to make your mcp server do the appropriate networking calls for you within its own process (just like a proxy mentioned by @shenli )

**Comment by andrewlayer** [comment]
So this actually is possible, but currently the most clients do not support HTTP transfer protocol.  This is an [example repository](https://github.com/andrewlayer/MCP-http-protocol) I put together that shows you how to create a server and use the inspector application to do it.  Again, Claude, Continue, and other do not support this yet, but I imagine it will come soon.

**Comment by friendlygeorge** [comment]
Hi — I have shipped a few npm-published MCP servers (`@supernova123/docker-mcp-server`, `@supernova123/system-monitoring-mcp-server`, `@supernova123/graphql-mcp-server`, `@supernova123/cron-scheduler-mcp-server`) and the hosting patterns that have worked for me in production setups:

1. **Run the stdio server on a jump host and tunnel.** `ssh -L` or Tailscale + the host invokes `npx -y @scope/server` locally — same UX as Claude Desktop but the server lives anywhere.

2. **Wrap with a thin HTTP shim.** Projects like `mcp-gateway` work, and so does running a stdio MCP server behind a Cloudflare Worker or FastAPI that speaks the Streamable HTTP transport.

3. **Package as a Docker container** so the binary does not need to be on the host PATH — that is why my Docker server uses a Dockerfile that bundles the stdio entrypoint. Easier to roll out via Compose or Swarm than a bare npx.

Until remote transports land broadly in clients, option (1) is honestly the lowest-friction path. The transport story is moving fast though — `@modelcontextprotocol/sdk` already has Streamable HTTP support, and the registry preview gives you a place to publish regardless of transport.

Hope this helps!

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
